In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv(r"C:\Users\pavan\OneDrive\Pictures\WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [5]:
#Top 10 Rows

df.head(10)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,No,Yes,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.4,No
7,6713-OKOMC,Female,0,No,No,10,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,No,Mailed check,29.75,301.9,No
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes
9,6388-TABGU,Male,0,No,Yes,62,Yes,No,DSL,Yes,...,No,No,No,No,One year,No,Bank transfer (automatic),56.15,3487.95,No


In [7]:
#The data information
#Columns and the number of entries with its data types 

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [18]:
#It indicates that even though Total charges is numerical it represented as categorical 
#So now we will convert into numerical and will do the analysis

df['TotalCharges'] = pd.to_numeric(df['TotalCharges'],errors='coerce')

In [20]:
#The numerical data summary

df.describe()

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000,7032.000000
mean,0.162147,32.371149,64.761692,2283.300441
std,0.368612,24.559481,30.090047,2266.771362
min,0.000000,0.000000,18.250000,18.800000
25%,0.000000,9.000000,35.500000,401.450000
50%,0.000000,29.000000,70.350000,1397.475000
75%,0.000000,55.000000,89.850000,3794.737500
max,1.000000,72.000000,118.750000,8684.800000


**Observations**

- Our data consists of a total 21 columns with 3 numerical and 18 categorical
- Our data consists of outliers but which doesn't affect the analysis as they represent the realistic values
- Our Target variable is Churn

In [22]:
#Checking if there any null values present

df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [28]:
#After we changed the datatype of the TotalCharges we found the null values

In [34]:
#Verifying if the customers with the null values in Total Charges are new

df[df['TotalCharges'].isnull()][['tenure','MonthlyCharges']]

,tenure,MonthlyCharges
488,0,52.55
753,0,20.25
936,0,80.85
1082,0,25.75
1340,0,56.05
3331,0,19.85
3826,0,25.35
4380,0,20.00
5218,0,19.70
6670,0,73.35


*Missing values in TotalCharges were due to new customers with zero tenure. Since they had not incurred any charges yet, I replaced those values with 0 to maintain data consistency*

In [39]:
df['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [41]:
df['Churn'].value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

*Approximately 26.53% of customers have churned, while 73.46% have retained the service. This indicates a moderate churn rate, highlighting the importance of identifying factors contributing to customer attrition.*

🔹**Gender vs Churn**

❓ **Question**

Does gender influence customer churn?

In [71]:
df.groupby('gender')['Churn'].value_counts(normalize=True).unstack()

Churn,No,Yes
gender,,
Female,0.730791,0.269209
Male,0.738397,0.261603


📊 **Insight**

The churn rate across genders shows minimal variation.

🧠 **Interpretation**

This suggests that gender does not significantly impact customer retention behavior.

✅ **Conclusion**

Gender is a weak predictor of churn and does not play a major role in customer attrition.

🔹**Contract vs Churn**

❓ **Question**

Does contract type influence churn?

In [69]:
df.groupby('Contract')['Churn'].value_counts(normalize=True).unstack()

Churn,No,Yes
Contract,,
Month-to-month,0.572903,0.427097
One year,0.887305,0.112695
Two year,0.971681,0.028319


📊 **Insight**

Customers with month-to-month contracts have significantly higher churn rates compared to those with long-term contracts.

🧠 **Interpretation**

This indicates that customers with short-term commitments are more likely to discontinue the service.

✅ **Conclusion**

Contract type is a strong factor influencing churn, with long-term contracts improving customer retention.

🔹**Tenure vs Churn**

❓ **Question**

Does tenure influence churn?


In [105]:
df.groupby('Churn')['tenure'].mean()

Churn
No     37.569965
Yes    17.979133
Name: tenure, dtype: float64

📊 **Insight**

Customers who churn have lower average tenure compared to those who remain.

🧠 **Interpretation**

This suggests that newer customers are more likely to discontinue the service.

✅ **Conclusion**

Tenure plays a significant role in churn, with shorter tenure associated with higher churn rates.

🔹**Monthly Charges vs Churn**

❓ **Question**

Does Monthly Charges influence churn?

In [86]:
df.groupby('Churn')['MonthlyCharges'].mean()

Churn
No     61.265124
Yes    74.441332
Name: MonthlyCharges, dtype: float64

📊 **Insight**

Customers who churn have higher average monthly charges compared to those who remain.

🧠 **Interpretation**

This suggests that higher pricing may contribute to customer dissatisfaction, increasing the likelihood of churn.

✅ **Conclusion**

Monthly charges appear to influence churn behavior, indicating that pricing strategy may play a role in customer retention.

🔹 **Final Observations**

- Gender shows negligible difference in churn, indicating it is not a significant factor.
- Contract analysis indicates that customers with long-term contracts have lower churn compared to month-to-month customers, making it a key factor.
- Tenure analysis shows that customers with shorter tenure are more likely to churn, while long-term customers tend to stay, making tenure an important factor.
- Monthly charges analysis indicates that higher charges are associated with increased churn, making it a moderate influencing factor.


Overall, churn is primarily driven by contract type, tenure, and pricing factors, while demographic attributes such as gender have minimal impact.